File: docker/sandbox/python/runner.py
import sys
import json
import time
import tracemalloc
import math


def normalize_json_value(value):
    if value is None or isinstance(value, (str, int, bool)):
        return value
    if isinstance(value, float):
        return value if math.isfinite(value) else str(value)
    if isinstance(value, dict):
        return {str(key): normalize_json_value(val) for key, val in value.items()}
    if isinstance(value, (list, tuple)):
        return [normalize_json_value(item) for item in value]
    return str(value)

def execute_solution(namespace, args):
    """
    Measure only the user function call window.
    Excludes parse/compile/bootstrap overhead from memory/time measurement.
    """
    tracemalloc.start()
    start_ns = time.perf_counter_ns()
    try:
        result = namespace["solution"](*args)
        error = None
        output = normalize_json_value(result)
    except Exception as e:
        output = None
        error = f"RUNTIME_ERROR: {e}"
    finally:
        elapsed_ms = (time.perf_counter_ns() - start_ns) / 1_000_000.0
        _, peak_bytes = tracemalloc.get_traced_memory()
        tracemalloc.stop()

    return {
        "output": output,
        "error": error,
        "timeMs": elapsed_ms,
        "memoryMb": peak_bytes / (1024 * 1024),
    }

def main():
    try:
        raw = sys.stdin.read()
        payload = json.loads(raw)
    except Exception as e:
        print(json.dumps({"output": None, "error": f"INTERNAL_ERROR: failed to parse input: {e}", "timeMs": 0.0, "memoryMb": 0.0}))
        sys.exit(1)

    code = payload.get("code", "")
    args = payload.get("args", [])

    namespace = {}
    try:
        exec(compile(code, "<solution>", "exec"), namespace)
    except SyntaxError as e:
        print(json.dumps({"output": None, "error": f"COMPILE_ERROR: {e}", "timeMs": 0.0, "memoryMb": 0.0}))
        sys.exit(0)
    except Exception as e:
        print(json.dumps({"output": None, "error": f"RUNTIME_ERROR: {e}", "timeMs": 0.0, "memoryMb": 0.0}))
        sys.exit(0)

    if "solution" not in namespace or not callable(namespace["solution"]):
        print(json.dumps({"output": None, "error": "RUNTIME_ERROR: 'solution' function not defined", "timeMs": 0.0, "memoryMb": 0.0}))
        sys.exit(0)

    print(json.dumps(execute_solution(namespace, args)))

if __name__ == "__main__":
    main()

File: docker/sandbox/dart/runner.dart
import 'dart:io';
import 'dart:convert';

String _toLiteral(dynamic a) {
  if (a == null) return 'null';
  if (a is String) return "'${a.replaceAll(r'\', r'\\').replaceAll("'", r"\'")}'" ;
  if (a is List) return '[${a.map(_toLiteral).join(', ')}]';
  if (a is Map) return '{${a.entries.map((e) => "'${e.key}': ${_toLiteral(e.value)}").join(', ')}}';
  return a.toString();
}

String toArgsLiteral(List<dynamic> args) => args.map(_toLiteral).join(', ');

void main() async {
  String raw;
  try {
    raw = await stdin.transform(utf8.decoder).join();
  } catch (e) {
    print(jsonEncode({'output': null, 'error': 'INTERNAL_ERROR: failed to read stdin: $e', 'timeMs': 0.0, 'memoryMb': 0.0}));
    exit(1);
  }

  Map<String, dynamic> payload;
  try {
    payload = jsonDecode(raw) as Map<String, dynamic>;
  } catch (e) {
    print(jsonEncode({'output': null, 'error': 'INTERNAL_ERROR: failed to parse input: $e', 'timeMs': 0.0, 'memoryMb': 0.0}));
    exit(1);
  }

  final code = payload['code'] as String? ?? '';
  final args = (payload['args'] as List<dynamic>?) ?? [];

  final tmpDir = await Directory('/tmp').createTemp('judge_');
  final sourceFile = File('${tmpDir.path}/solution.dart');
  final resultFile = File('${tmpDir.path}/result.txt');

  final argsLiteral = toArgsLiteral(args);

  // Escape single quotes in the result file path for embedding in Dart string literal
  final safeResultPath = resultFile.path.replaceAll(r'\', r'\\').replaceAll("'", r"\'");

  // Generate solution.dart: user code + main() that writes result to a file
  // Uses string concatenation to avoid Dart template interpolation conflicts
  await sourceFile.writeAsString(
    "import 'dart:convert';\n" +
    "import 'dart:io';\n\n" +
    "Object? __judgeNormalizeJsonValue(Object? value) {\n" +
    "  if (value == null || value is num || value is bool || value is String) return value;\n" +
    "  if (value is List) return value.map((item) => __judgeNormalizeJsonValue(item)).toList();\n" +
    "  if (value is Map) return { for (final entry in value.entries) entry.key.toString(): __judgeNormalizeJsonValue(entry.value) };\n" +
    "  return value.toString();\n" +
    "}\n\n" +
    code + "\n\n" +
    "void main() {\n" +
    "  final resultFile = File('$safeResultPath');\n" +
    "  final sw = Stopwatch()..start();\n" +
    "  try {\n" +
    "    final result = solution($argsLiteral);\n" +
    "    sw.stop();\n" +
    "    resultFile.writeAsStringSync(jsonEncode({'status': 'OK', 'output': __judgeNormalizeJsonValue(result), 'timeMs': sw.elapsedMicroseconds / 1000.0}));\n" +
    "  } catch (e) {\n" +
    "    sw.stop();\n" +
    "    resultFile.writeAsStringSync(jsonEncode({'status': 'RUNTIME_ERROR', 'error': e.toString(), 'timeMs': sw.elapsedMicroseconds / 1000.0}));\n" +
    "  }\n" +
    "}\n"
  );

  // Run with `dart run` (JIT — no native binary written, safe with noexec tmpfs)
  final result = await Process.run('dart', ['run', sourceFile.path]);

  String? errorMsg;
  Object? outputVal;
  double timeMs = 0.0;

  if (resultFile.existsSync()) {
    final resultPayload = jsonDecode(resultFile.readAsStringSync()) as Map<String, dynamic>;
    final status = resultPayload['status'] as String? ?? 'RUNTIME_ERROR';
    timeMs = (resultPayload['timeMs'] as num?)?.toDouble() ?? 0.0;

    if (status == 'OK') {
      outputVal = resultPayload['output'];
    } else {
      errorMsg = '$status: ${resultPayload['error'] ?? ''}';
    }
  } else {
    // result file not written → compile/syntax error
    final stderr = (result.stderr as String).replaceAll('\n', ' ').trim();
    errorMsg = 'COMPILE_ERROR: $stderr';
  }

  await tmpDir.delete(recursive: true);

  if (errorMsg != null) {
    print(jsonEncode({'output': null, 'error': errorMsg, 'timeMs': timeMs, 'memoryMb': 0.0}));
  } else {
    print(jsonEncode({'output': outputVal, 'error': null, 'timeMs': timeMs, 'memoryMb': 0.0}));
  }
}
